### PR TITLE
HTML Games

### PR DESCRIPTION
**Pull request by Mrmaxmeier**
This PR adds the ability to download and launch web games (#47).
Launching a web game will spin up a simple http server on a random port and open a new BrowserWindow inside of electron. Closing the game / window will shut the http server down.
The dependencies `serve-static` and `finalhandler` are used to simplify the http server.

:construction: To-dos
- [x] no way for games to store gamesaves / persistent data
- [x] split launch.js
- [x] handle games that have all content in a subdir
- [x] tests
- [ ] potential security problems
- [x] fetch window size (leafo/itch.io#310)
- [x] don't use hacky `__proto__` in [tests](https://github.com/itchio/itch/pull/328/files#diff-53bacd6fd41cd0c4417df8e032b0ef40R28)

### CODE REVIEW

**Comment by fasterthanlime** [comment]
0.12.0 is in feature freeze so I'm not going to merge it before then — but thanks for opening the PR early! :+1:

**Comment by fasterthanlime** [comment]
As far as gamesaves go, you can use a `persist:game_id` partition, cf. the session docs: http://electron.atom.io/docs/v0.36.0/api/session/

File: app/tasks/launch.js [hunk 1]
```diff
@@ -105,6 +114,46 @@ let self = {
     return self.launch(exe_path, [], opts)
   },
 
+  launch_html_cave: async function(opts, cave) {
```
Comment (fasterthanlime) [inline app/tasks/launch.js]: I would probably feel better about having this in another module — launch already does so much.

I'm not sure what the right granularity would be, maybe `launch.js`, `launch/native.js` and `launch/html.js` ?

File: app/tasks/launch.js [hunk 2]
```diff
@@ -105,6 +114,46 @@ let self = {
...
+    let game = await db.find_one({_table: 'games', id: cave.game_id})
+    let app_path = CaveStore.app_path(cave.install_location, opts.id)
+    if (!fs.existsSync(path.join(app_path, 'index.html'))) {
+      throw new Error('html game missing index.html')
+    }
+    let win = new BrowserWindow({
+      title: game.title,
+      icon: './static/images/itchio-tray-x4.png',
+      width: 1280, height: 720,
```
Comment (fasterthanlime) [inline app/tasks/launch.js]: games can specify actual width/height but we don't have that info yet, just opened https://github.com/leafo/itch.io/issues/310

File: app/tasks/launch.js [hunk 3]
```diff
@@ -105,6 +114,46 @@ let self = {
...
+    server.on('listening', function() {
+      port = server.address().port
+      log(opts, `serving game on port ${port}`)
+      win.loadURL(`http://localhost:${port}/index.html`)
+      if (process.env.DEVTOOLS === '1') {
+        win.webContents.openDevTools({detach: true})
+      }
+    })
+
+    return new Promise((resolve, reject) => {
```
Comment (fasterthanlime) [inline app/tasks/launch.js]: returning a promise is now an anti-pattern — instead, you can await one.

example:

``` javascript
async function waste_time () {
  let p = new Promise((resolve, reject) => {
    setTimeout(resolve, 1000)
  })
  console.log(`Waiting one second...`)
  await p
  console.log(`Hey, there we are!`)  
}
```

File: app/tasks/launch.js [hunk 4]
```diff
@@ -113,6 +162,14 @@ let self = {
     let id = opts.id
 
     let cave = await CaveStore.find(id)
+    if (!cave.uploads) {
+      throw new Error('need cached uploads')
+    }
+    let has_native = _.values(cave.uploads).filter((upload) => !!upload[`p_${os.itch_platform()}`]).length > 0
```
Comment (fasterthanlime) [inline app/tasks/launch.js]: I'm sure underscore has a better primitive than 'filter' for that

**Comment by Mrmaxmeier** [comment]
@fasterthanlime Thanks for your review, I'll rebase after 0.12.0 is out.

**Comment by Mrmaxmeier** [comment]
@fasterthanlime I've rebased and haven't found any broken html games yet. The code is ready to be reviewed and/or merged.

File: appsrc/tasks/find-upload.js [hunk 5]
```diff
@@ -48,6 +48,12 @@ let self = {
       score -= 100
     }
 
+    /* Native uploads are preferred */
+    if (upload.type === 'html') {
+      score -= 20
+      score -= 20
```
Comment (fasterthanlime) [inline appsrc/tasks/find-upload.js:54]: merge glitch or did you mean to subtract 40 ?

Comment (Mrmaxmeier) [inline appsrc/tasks/find-upload.js:54]: that wasn't intended. the git history for this pr is pretty bad. i'll push a fix and remove [this commit](https://github.com/Mrmaxmeier/itch/commit/6aabc8072ec7e8c260923b6c85a21b25d37a170f).

Comment (fasterthanlime) [inline appsrc/tasks/find-upload.js:54]: don't worry, I'll fix it during review!

**Comment by fasterthanlime** [comment]
I'm still doing a few changes to help debug some games (Snow World for example doesn't load for me) but overall it looks good.

Great job @Mrmaxmeier, thanks for your patience! :stars:

**Comment by fasterthanlime** [comment]
Just removed `__proto__` usage in master by using a separate http-server module.

**Comment by fasterthanlime** [comment]
(Note: I've also tried registering a custom protocol in Electron to avoid having to spin up a real http server but it didn't go well. The custom protocol would be available in the main app window but not in BrowserWindows created afterwards - requests would fail with net::ErrUnknownProtocol I'm not sure if it's a bug on their side or what but it seems spinning an http server is the most reliable for now.).

**Pull request merged**